io-amqp1_0-impl/src/main/java/org/apache/pulsar/ecosystem/io/amqp/AmqpSource.java

```diff
@@ -111,17 +114,41 @@ public void onMessage(Message message) {
     private static class QpidJmsRecord implements Record<ByteBuffer> {
         private final Optional<String> key;
         private final ByteBuffer value;
+        private final Message jmsMessage;
 
         @Override
         public Schema<ByteBuffer> getSchema() {
             return Schema.BYTEBUFFER;
         }
+
+        @Override
+        public void ack() {
+            try {
+                jmsMessage.acknowledge();
+            } catch (JMSException e) {
+                log.error("Failed to acknowledge qpid jms message.", e);
+            }
+        }
+
+        @Override
+        public void fail() {
+            try {
+                if (jmsMessage instanceof JmsMessage) {
```

Comment:
It seems that we always/only accept `JmsMessage` in `MessageListenerImpl#onMessage `. Perhaps we can assert the type there and use `JmsMessage` in `QpidJmsRecord`.

Reply:
Good find. We change it.